### PR TITLE
Discrete grid positioning, no overlap

### PR DESCRIPTION
**Pull request by Lev-Stambler**
- Create an algorithm to avoid overlap of new items on the grid. This is done by checking in a spiral with a growing radius centered by the original position.
- Create a discrete unit of positioning (0.5 km) so that all automatically placed items are either 0.5 km, 1km, 1.5km, etc apart
- Create an error if items cannot fit onto the grid
- Add tests

### CODE REVIEW

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts]: here and above, if the imports use everything in the file, feel free to use `import *` instead

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:306]: since xoffset and y offset start off as 0, won't this check the exact same position you already checked in line 283?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts]: can we shorten this test example at all in order to make it more readable + concise?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts:123]: this seems like two separate tests - (1) action = brain.computeAction and expect getSupplyingPaths length, and (2) grid.takeAction

Please split these up :)

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts:160]: nit: add extra line between test cases

**Review by vitashubin — CHANGES_REQUESTED** [review]
great job on this PR despite the complexity of the algorithm, very well done :)

**Review by cwlarson-google — APPROVED** [review]

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:306]: Oh yeah it will. I removed the if condition

File: biogrid/apps/api/src/app/services/Biogrid.spec.ts
Comment (vitashubin) [inline biogrid/apps/api/src/app/services/Biogrid.spec.ts]: nit: remove 'that' from test name

File: biogrid/apps/api/src/app/services/Biogrid.spec.ts
Comment (vitashubin) [inline biogrid/apps/api/src/app/services/Biogrid.spec.ts]: I mentioned this in another PR as well - why -1 and -2? 
can you do:
`const endDate = new Date(); const startDate = endDate.getDate() - 1;` instead?

File: biogrid/apps/api/src/app/services/Biogrid.spec.ts
Comment (vitashubin) [inline biogrid/apps/api/src/app/services/Biogrid.spec.ts]: const instead of let

File: biogrid/apps/api/src/app/services/Biogrid.ts
Comment (vitashubin) [inline biogrid/apps/api/src/app/services/Biogrid.ts]: inline comment on what 10 is or replace with constant name

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts]: I believe @cwlarson-google mentioned this in another pr - for constructors that take a lot of parameters, consider using a single object parameter instead

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:307]: super nit: all your code uses >=4 instead of >3, can we update this comment to be consistent with the code?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts]: nit: `minumum dsitance of ${GRID_DISTANCES } km apart`

**Review by vitashubin — CHANGES_REQUESTED** [review]

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts]: Fixed!

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts]: I don't think we want default params here. TS will check that all the necessary params are present, and we want to have the user set all these values instead of allowing them to default to an empty string id or other default values
@cwlarson-google feel free to correct me if I'm wrong here - I'm going off a comment you made on a separate PR

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts]: I know this isn't part of your PR but why is it lessthanorequal rather than exactly the expected value?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:165]: also not part of your PR, but what do you mean by "oldGridItem" here?

**Review by vitashubin — CHANGES_REQUESTED** [review]

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts
Comment (cwlarson-google) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts]: If you have optional params, the "standard" way that I've seen (could be something my team made up, could be a real standard) is to set the default when declaring the class property and then conditionally overwriting it if a value is passed in:

```
export class BioBattery implements Battery {
  private energyInJoules: Energy = 100;    // Defaults to 100 Energy

  constructor(params: BatteryParams) {
    if (params.energyInJoules) {
      this.energyInJoules = params.energyInJoules;
    }
  }
}
```

Since you need to run things through a validator, this might make things more complicated?

At the end of the day, I think consistency is more important than anything else.  If this is how the whole app is approaching defaults, then fine

Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery-class.ts]: Ok sure, it should be standard now

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts
Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts]: @RolandNaijuka, correct me if I am wrong, I believe it has to do with efficiency and losses from delivering the energy. So, not all outputed energy is delivered

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (Lev-Stambler) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:165]: @RolandNaijuka, also correct me if I am wrong, I believe it has to do with the fact that the grid item gets updated?

Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:165]: that is right, this gridItem is updated at the end of the for loop in the clonedGraph

File: biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts
Comment (RolandNaijuka) [reply biogrid/libs/biogrid-simulator/src/lib/biobattery/biobattery.spec.ts]: That is right @Lev-Stambler

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts]: looks like the solar panels are created now, can you remove this comment?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts
Comment (vitashubin) [reply biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts:307]: I don't see this updated :)

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts]: nit, can you add an inline comment on what `[]` is as well?

File: biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts
Comment (vitashubin) [inline biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid.spec.ts]: `[]` here as well

**Review by vitashubin — APPROVED** [review]
Approved pending a few small comments - please make sure to address them before merging (they should be very quick)

**Comment by Lev-Stambler** [comment]
@vitashubin the newPos was updated by removing the if statement

**Pull request merged**
